File: app/javascript/channels/locations_channel.js
import consumer from "./consumer"

document.addEventListener('turbolinks:load', () => {

  const task = document.querySelector('section[class~="task-show-container"]');

  // 進到 task show 頁面才可以得到 task 的 id，會比對後端打過來的 employee_id，一致才更新位置資訊（預設 employee 開啟 task 的 show 頁面就傳遞一次位置資訊給 employer）
  if (task) {

    // document.getElementById("report_location").addEventListener('onclick', () => {
    //   console.log("testing")
    // });

    const employee_id = task.dataset.employeeId
    const employer_id = task.dataset.employerId
  
    // 要寫一個確定 employee 跟 employer match 的判斷

    // employer is the subscriber and employee broadcast location to employer
    // the following means consumer (employer) subscribes the channel of employee
    const locationChannel = consumer.subscriptions.create({ channel: "LocationsChannel", employee_id: employee_id }, {
      
      connected() {
        // Called when the subscription is ready for use on the server
        console.log("監看 employee 中")
      },
  
      disconnected() {
        // Called when the subscription has been terminated by the server
      },
  
      received(data) {
        console.log("==============================")
        console.log("==============================")
        console.log(data)
        // Called when there's incoming data on the websocket for this channel
        // console.log(data)
      }
    });
    
  }

})

File: app/channels/locations_channel.rb
class LocationsChannel < ApplicationCable::Channel

  # 講一下邏輯
  # 在前端，employee 點回報位置按鈕，JS 會 call google map 的 callback，等到 google map 回傳 emplyee 位置
  # 後透過 actioncable JS 裡定義的 speak 將資料傳出去，至於要傳去哪裡？
  # subscribed 意思為用戶訂閱了哪些頻道，而我特意寫死在 params[:employeeId]，employeeId 是前端的
  # consumer.subscriptions.create({ channel: "LocationsChannel", employeeId: employeeId } 裡的 emplyeeId
  # 所以寫死後，speak 方法裡的 broadcast 就只能 broadcast 到這個 room
  # 接下來就是在前端判斷是不是該這個雇主接這個應徵者傳出的位置資訊

  def subscribed
    stream_from params[:employeeId]
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    # broadcast the message to received function of room_channel
    ActionCable.server.broadcast(params[:employeeId], { location: data["location"] })
  end
  
end
